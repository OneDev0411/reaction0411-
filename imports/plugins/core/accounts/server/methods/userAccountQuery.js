import { Accounts } from "/lib/collections";
import { Reaction } from "/lib/api";

export default function userAccountQuery(id) {
  // Check to make sure current user has permissions to view queried user
  if (!Reaction.hasPermission("reaction-accounts")) throw new Error("User does not have permission");

  // Query the accounts collection to find user by ID
  const userAccount = Accounts.findOne({
    _id: id
  });

  // If user is not found, throw an error
  if (!userAccount) throw new Error("No account found");

  // If account is found, return userAccount
  return userAccount;
}
