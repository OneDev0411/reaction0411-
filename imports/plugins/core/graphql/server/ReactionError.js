import ExtendableError from "es6-error";

export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
  }
}
