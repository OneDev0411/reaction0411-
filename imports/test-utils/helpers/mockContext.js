const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
