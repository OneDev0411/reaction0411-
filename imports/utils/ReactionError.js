import ExtendableError from "es6-error";

/**
 * @name ReactionError
 * @class
 * @memberof GraphQL
 * @summary A type of error that allows additional details to be provided, which can
 *   then be sent back to the client for a GraphQL request.
 */
export default class ReactionError extends ExtendableError {
  constructor(error, message, eventData = {}) {
    super(message);
    this.eventData = eventData;
    this.error = error;
    // Newer versions of DDP use this property to signify that an error
    // can be sent back and reconstructed on the calling client.
    this.isClientSafe = true;
    // DDP expects this to be in a property named `reason`
    this.reason = message;
    // DDP expects this to be in a property named `details`
    this.details = eventData;
  }

  clone() {
    return new ReactionError(this.error, this.reason, this.eventData);
  }
}
