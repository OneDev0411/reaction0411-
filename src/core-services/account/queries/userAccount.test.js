import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import ReactionError from "@reactioncommerce/reaction-error";
import userAccountQuery from "./userAccount.js";

const fakeAccountId = "FAKE_USER_ID_FOR_QUERY";
const fakeAccount = { _id: fakeAccountId, userId: fakeAccountId, shopId: "ACCOUNT_SHOP_ID" };
const fakeContextAccount = { _id: mockContext.accountId, userId: mockContext.userId, shopId: "ACCOUNT_SHOP_ID" };

beforeEach(() => {
  jest.resetAllMocks();
});

test("throws if account not found", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);
  await expect(userAccountQuery(mockContext, fakeAccountId)).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.collections.Accounts.findOne).toHaveBeenCalledWith({ _id: fakeAccountId });
});

test("throws if userHasPermission returns false and the user ID is not the context user ID", async () => {
  mockContext.validatePermissions.mockImplementation(() => {
    throw new ReactionError("access-denied", "Access Denied");
  });

  mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
  await expect(userAccountQuery(mockContext, fakeAccountId)).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.validatePermissions).toHaveBeenCalledWith(
    "reaction:legacy:accounts",
    "read",
    { owner: fakeAccountId }
  );
});

test("returns the account without calling userHasPermission if the user ID is the context user ID", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeContextAccount);
  const result = await userAccountQuery(mockContext, mockContext.accountId);
  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
  expect(result).toEqual(fakeContextAccount);
});

test("returns the account if the user ID is not the context user ID but userHasPermission returns true", async () => {
  mockContext.validatePermissions.mockReturnValueOnce(Promise.resolve(null));
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
  const result = await userAccountQuery(mockContext, fakeAccountId);
  expect(mockContext.validatePermissions).toHaveBeenCalledWith(
    "reaction:legacy:accounts",
    "read",
    { owner: fakeAccountId }
  );
  expect(result).toEqual(fakeAccount);
});
