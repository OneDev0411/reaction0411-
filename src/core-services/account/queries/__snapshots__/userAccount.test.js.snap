// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if account not found 1`] = `"No account found"`;

exports[`throws if userHasPermissionLegacy returns false and the user ID is not the context user ID 1`] = `"Access Denied"`;
